vector/src/main/java/im/vector/app/features/analytics/DecryptionFailureTracker.kt

```diff
@@ -16,149 +16,278 @@
-                    .forEach { failedEventId ->
-                        analyticsTracker.capture(Error(
-                                context = aggregation.key.first,
-                                domain = Error.Domain.E2EE,
-                                name = aggregation.key.second,
-                                cryptoModule = currentModule
-                        ))
-                        alreadyReported.add(failedEventId)
-                    }
+    private fun handleEventDecrypted(eventId: String) {
+        Timber.v("Handle event decrypted $eventId time: ${clock.epochMillis()}")
+        // Only consider if it was tracked as a failure
+        val trackedFailure = trackedEventsMap[eventId] ?: return
+
+        // Grace event if decrypted under 4s
+        val now = clock.epochMillis()
+        val timeToDecrypt = now - trackedFailure.timeStamp
+        Timber.v("Handle event decrypted timeToDecrypt: $timeToDecrypt for event $eventId")
+        if (timeToDecrypt < GRACE_PERIOD_MILLIS) {
+            Timber.v("Grace event $eventId")
+            trackedEventsMap.remove(eventId)
+            return
+        }
+        // We still want to report but with the time it took
+        if (trackedFailure.timeToDecryptMillis == null) {
+            val decryptionFailure = trackedFailure.copy(timeToDecryptMillis = timeToDecrypt)
+            trackedEventsMap[eventId] = decryptionFailure
+            reportFailure(decryptionFailure)
         }
     }
 
-    private fun MXCryptoError.ErrorType.toAnalyticsErrorName(): DetailedErrorName {
-        val detailed = "$name | mxc_crypto_error_type"
-        val errorName = when (this) {
-            MXCryptoError.ErrorType.UNKNOWN_INBOUND_SESSION_ID,
-            MXCryptoError.ErrorType.KEYS_WITHHELD -> Error.Name.OlmKeysNotSentError
-            MXCryptoError.ErrorType.OLM -> Error.Name.OlmUnspecifiedError
-            MXCryptoError.ErrorType.UNKNOWN_MESSAGE_INDEX -> Error.Name.OlmIndexError
-            else -> Error.Name.UnknownError
+     fun utdDisplayedInTimeline(event: TimelineEvent) {
+        post {
+            // should be tracked (unless already reported)
+            val eventId = event.root.eventId ?: return@post
+            val trackedEvent = trackedEventsMap[eventId] ?: return@post
+
+            trackedEventsMap[eventId] = trackedEvent.copy(wasVisibleOnScreen = true)
+        }
+    }
+
+    // This will mutate the trackedEventsMap, so don't call it while iterating on it.
+    private fun reportFailure(decryptionFailure: DecryptionFailure) {
+        Timber.v("Report failure for event ${decryptionFailure.failedEventId}")
+        val error = decryptionFailure.toAnalyticsEvent()
+
+        analyticsTracker.capture(error)
+
+        // now remove from tracked
+        trackedEventsMap.remove(decryptionFailure.failedEventId)
+        // mark as already reported
+        alreadyReported.add(decryptionFailure.failedEventId)
```

Comment:
`alreadyReported` does not seem to be read. 
It was used to filter the Events previously (line 140 in previous version)

Reply:
Oh good catch. I fixed the unit test that was not correctly testing that